File: frontend/src/app/modules/settings/services/player-flight-assignment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Player } from './player.service';

export interface PlayerFlightAssignment {
  id?: string;  // GUID compatibility
  playerId: string;  // GUID compatibility
  flightId: string;  // GUID compatibility
  seasonId: string;  // GUID compatibility
  sessionStartWeekNumber: number;
  isFlightLeader: boolean;
  handicapAtAssignment?: number;
  assignmentDate?: string;
  player?: Player; // For joined data
}

@Injectable({
  providedIn: 'root'
})
export class PlayerFlightAssignmentService {
  private readonly apiUrl = '/api/player-flight-assignments';

  constructor(private http: HttpClient) { }

  getAssignmentsByFlight(flightId: string): Observable<PlayerFlightAssignment[]> {
    return this.http.get<PlayerFlightAssignment[]>(`${this.apiUrl}/flight/${flightId}`);
  }

  getAssignmentsByFlightAndSession(flightId: string, seasonId: string, sessionStartWeekNumber: number): Observable<PlayerFlightAssignment[]> {
    const params = new HttpParams()
      .set('seasonId', seasonId)
      .set('sessionStartWeekNumber', sessionStartWeekNumber.toString());
    return this.http.get<PlayerFlightAssignment[]>(`${this.apiUrl}/flight/${flightId}/session`, { params });
  }

  getAssignmentsByPlayerAndSeason(playerId: string, seasonId: string): Observable<PlayerFlightAssignment[]> {
    return this.http.get<PlayerFlightAssignment[]>(`${this.apiUrl}/player/${playerId}/season/${seasonId}`);
  }

  getAssignmentsBySession(seasonId: string, sessionStartWeekNumber: number): Observable<PlayerFlightAssignment[]> {
    const params = new HttpParams()
      .set('seasonId', seasonId)
      .set('sessionStartWeekNumber', sessionStartWeekNumber.toString());
    return this.http.get<PlayerFlightAssignment[]>(`${this.apiUrl}/session`, { params });
  }

  getPlayerAssignmentForSession(playerId: string, seasonId: string, sessionStartWeekNumber: number): Observable<PlayerFlightAssignment> {
    const params = new HttpParams()
      .set('seasonId', seasonId)
      .set('sessionStartWeekNumber', sessionStartWeekNumber.toString());
    return this.http.get<PlayerFlightAssignment>(`${this.apiUrl}/player/${playerId}/session`, { params });
  }

  addAssignment(assignment: PlayerFlightAssignment): Observable<PlayerFlightAssignment> {
    return this.http.post<PlayerFlightAssignment>(this.apiUrl, assignment);
  }

  removeAssignment(assignmentId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${assignmentId}`);
  }

  updateAssignment(assignment: PlayerFlightAssignment): Observable<PlayerFlightAssignment> {
    return this.http.put<PlayerFlightAssignment>(`${this.apiUrl}/${assignment.id}`, assignment);
  }

  getAssignmentsByPlayer(playerId: string): Observable<PlayerFlightAssignment[]> {
    return this.http.get<PlayerFlightAssignment[]>(`${this.apiUrl}/player/${playerId}`);
  }

  getAllAssignments(): Observable<PlayerFlightAssignment[]> {
    return this.http.get<PlayerFlightAssignment[]>(this.apiUrl);
  }
}

File: backend/Business/PlayerFlightAssignmentService.cs
namespace GolfLeagueManager
{
    public class PlayerFlightAssignmentService
    {
        private readonly IPlayerFlightAssignmentRepository _repository;

        public PlayerFlightAssignmentService(IPlayerFlightAssignmentRepository repository)
        {
            _repository = repository;
        }

        public void AddAssignment(PlayerFlightAssignment assignment)
        {
            if (assignment.Id == Guid.Empty)
            {
                assignment.Id = Guid.NewGuid();
            }
            ValidateAssignment(assignment);
            _repository.AddAssignment(assignment);
        }

        public IEnumerable<PlayerFlightAssignment> GetAssignmentsByFlight(Guid flightId)
        {
            return _repository.GetAssignmentsByFlight(flightId);
        }

        public IEnumerable<PlayerFlightAssignment> GetAssignmentsByFlightAndSession(Guid flightId, Guid seasonId, int sessionStartWeekNumber)
        {
            return _repository.GetAssignmentsByFlightAndSession(flightId, seasonId, sessionStartWeekNumber);
        }

        public IEnumerable<PlayerFlightAssignment> GetAssignmentsByPlayer(Guid playerId)
        {
            return _repository.GetAssignmentsByPlayer(playerId);
        }

        public IEnumerable<PlayerFlightAssignment> GetAssignmentsByPlayerAndSeason(Guid playerId, Guid seasonId)
        {
            return _repository.GetAssignmentsByPlayerAndSeason(playerId, seasonId);
        }

        public IEnumerable<PlayerFlightAssignment> GetAssignmentsBySession(Guid seasonId, int sessionStartWeekNumber)
        {
            return _repository.GetAssignmentsBySession(seasonId, sessionStartWeekNumber);
        }

        public PlayerFlightAssignment? GetPlayerAssignmentForSession(Guid playerId, Guid seasonId, int sessionStartWeekNumber)
        {
            return _repository.GetPlayerAssignmentForSession(playerId, seasonId, sessionStartWeekNumber);
        }

        public PlayerFlightAssignment? GetAssignmentById(Guid id)
        {
            return _repository.GetAssignmentById(id);
        }

        public bool UpdateAssignment(PlayerFlightAssignment assignment)
        {
            ValidateAssignment(assignment);
            return _repository.UpdateAssignment(assignment);
        }

        public bool RemoveAssignment(Guid id)
        {
            return _repository.RemoveAssignment(id);
        }

        public IEnumerable<PlayerFlightAssignment> GetAllAssignments()
        {
            return _repository.GetAllAssignments();
        }

        private void ValidateAssignment(PlayerFlightAssignment assignment)
        {
            if (assignment.PlayerId == Guid.Empty)
                throw new ArgumentException("Player ID is required.");

            if (assignment.FlightId == Guid.Empty)
                throw new ArgumentException("Flight ID is required.");

            if (assignment.SeasonId == Guid.Empty)
                throw new ArgumentException("Season ID is required.");

            if (assignment.SessionStartWeekNumber <= 0)
                throw new ArgumentException("Session start week number must be greater than 0.");

            if (assignment.HandicapAtAssignment < 0)
                throw new ArgumentException("Handicap cannot be negative.");
        }
    }
}
